feat: trigger alignWallets when  basicFunctionality ON

fabiobozzo opened the pull request:
## **Description**

Trigger multichain account alignment when the "Basic Functionality" toggle is enabled in MetaMask.

When users enable the "Basic Functionality" setting, this PR automatically calls `MultichainAccountService:setBasicFunctionality()` which handles both provider state management and multichain account alignment in a single operation. This ensures that multichain account groups are properly aligned with accounts for all available providers when external services are enabled.

## **Changelog**

* Add actions to PreferencesController's allowed actions (multichain builds only):
  * `MultichainAccountService:setBasicFunctionality` 
  * `MultichainAccountService:alignWallets`
* Call `setBasicFunctionality` when basic functionality toggle changes (handles both provider state and alignment)
* Add comprehensive test coverage for service integration

## **Related issues**

[MUL-343](https://consensyssoftware.atlassian.net/browse/MUL-343)

## **Manual testing steps**

1. Open MetaMask extension (multichain build)
2. Go to Settings > Advanced > Basic Functionality  
3. Toggle "Basic Functionality" from OFF to ON
4. Check browser console - should see `setBasicFunctionality` call with alignment activity
5. Verify that multichain account groups are aligned with all available providers
6. In non-multichain builds: Toggle should work without multichain service calls

## **Screenshots/Recordings**

<img width="1628" height="563" alt="image" src="https://github.com/user-attachments/assets/144d0fdc-1dcd-42c6-a05e-6b2fdeaa3810" />


### **Before**
- Basic functionality toggle had no effect on multichain account alignment
- Account groups could be missing accounts for certain providers

### **After** 
- Enabling basic functionality automatically aligns all multichain account groups
- Console shows alignment activity when toggle is changed

[MUL-343]: https://consensyssoftware.atlassian.net/browse/MUL-343?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
feat(multichain-accounts): trigger alignWallets on basicFunctionality turned on

Commit:
feat(multichain-accounts): trigger alignWallets on basicFunctionality turned on

Commit:
refactor(multichain): use setBasicFunctionality action MultichainAccountService instead of alignWallets directly

Commit:
refactor(multichain): simplify handler signature for setBasicFunctionality action

Commit:
fix lint issues

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
chore: bump multichain-account-service version to preview release

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
add MultichainAccountService to e2e test fixtures

Commit:
Trigger Build

Commit:
address cursorbot pr review

Commit:
Trigger Build

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

metamaskbot committed:
Update LavaMoat policies

Commit:
feat: add MultichainAccountService to error state snapshots

Commit:
feat: add MultichainAccountService to error state snapshots

Commit:
update memStateKey in MultichainAccountServiceInit

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
adjusted state snapshot JSON files to ensure consistent representation of MultichainAccountService across error states.

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

metamaskbot committed:
Update LavaMoat policies

Commit:
try and make e2e tests pass

Commit:
try and make e2e tests pass

Commit:
Enhance e2e test for Snap revoke permission with improved error handling and retry logic

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
Refactor e2e test for Snap revoke permission to improve error handling and streamline dialog interactions

Commit:
debug flaky e2e test

Commit:
try and make e2e tests pass

Commit:
try and make e2e tests pass

Commit:
reset faulty e2e test

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
yarn dedupe

Commit:
fix import order

Commit:
bump multichain-account-service to version 0.5.0-preview

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

fabiobozzo commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
chore: bump dep multichain-account-service preview build

Commit:
Update app/scripts/controllers/preferences-controller.ts

Co-authored-by: Charly Chevalier <charly.chevalier@consensys.net>

Commit:
subscribe to PreferencesController state changes to trigger basic functionality in MultichainAccountService

fabiobozzo commented on app/scripts/controllers/preferences-controller.ts:
```diff
@@ -515,6 +532,22 @@ export class PreferencesController extends BaseController<
+          error,
+        );
+      });
+    ///: END:ONLY_INCLUDE_IF
```
Good one! 💯 Please check it again now

Commit:
prettify

ccharly commented on app/scripts/controllers/preferences-controller.test.ts:
```diff
@@ -63,6 +63,9 @@ const setupController = ({
         'AccountsController:getSelectedAccount',
         'AccountsController:setSelectedAccount',
         'NetworkController:getState',
+        ///: BEGIN:ONLY_INCLUDE_IF(multichain)
+        'MultichainAccountService:setBasicFunctionality',
+        ///: END:ONLY_INCLUDE_IF
```
Yes, indeed, we sometimes leave them there for "contexts", but as discussed internally, we should really cleanup those old code-fences at some point.

ccharly reviewed (changes requested):
We need to remove the preview package.

Commit:
bump deps to actual release and remove code fences

Commit:
merge main into feat branch and update dependencies

Commit:
address reviewer feedback: remove unnecessary code fences

- Remove code fences from test files (no effect on tests)
- Remove multichain code fences from controller (not needed)
- Keep preview package temporarily due to peer dependency conflicts

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
bump multichain-account-service to 0.5.0

Commit:
bump multichain-account-service to 0.5.0

Commit:
refactor: remove unnecessary multichain code fences per reviewer feedback

- Remove code fences around MultichainAccountService in e2e test fixtures
- Remove code fences around MultichainAccountService actions and subscription
- Extract MultichainAccountService from multichain controller blocks
- Keep legitimate multichain code fences for other features intact

Addresses gantunesr feedback about unnecessary code fences affecting test files
and controller logic.

fabiobozzo commented on package.json:
```diff
@@ -312,7 +312,7 @@
     "@metamask/message-signing-snap": "1.1.3",
     "@metamask/metamask-eth-abis": "^3.1.1",
-    "@metamask/multichain-account-service": "^0.4.0",
+    "@metamask/multichain-account-service": "npm:@metamask-previews/multichain-account-service@0.5.0-preview-970969ca",
```
^0.5.0 ✔️

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
test: add missing MultichainAccountService:alignWallets action to test setup

- Add 'MultichainAccountService:alignWallets' to allowedActions in preferences controller test
- Add corresponding mock handler for alignWallets action
- Ensures test controller matches production controller's allowed actions

Addresses cursor bot feedback about missing action in test controller.

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
fix: add defensive error handling for MultichainAccountService calls

- Wrap MultichainAccountService:setBasicFunctionality call in try-catch
- Handle case where service is not available in test environment
- Prevents test failures while maintaining production functionality
- Addresses CI test failures in metamask-controller.test.js

ccharly commented on app/scripts/controller-init/messengers/accounts/multichain-account-service-messenger.ts:
```diff
@@ -46,6 +48,7 @@ export function getMultichainAccountServiceMessenger(
       'KeyringController:stateChange',
       'AccountsController:accountAdded',
       'AccountsController:accountRemoved',
+      'PreferencesController:stateChange',
```
Not sure we need this actually, since we're handling the subscription to that event "externally".

Also, the `@metamask/preferences-controller` is not the same as the "local" preferences-controller IIRC.

So we can probably remove this.

fabiobozzo replied:
Absolutely, I double checked we don't need it.

Commit:
refactor: remove unnecessary PreferencesController:stateChange from MultichainAccountService messenger

- Remove PreferencesControllerStateChangeEvent import and type usage
- Remove 'PreferencesController:stateChange' from allowedEvents
- External subscription handled in metamask-controller.js is sufficient

Addresses ccharly feedback about redundant event permission.

ccharly commented on app/scripts/metamask-controller.js:
```diff
@@ -1857,6 +1859,38 @@ export default class MetamaskController extends EventEmitter {
+        if (prevUseExternalServices !== currUseExternalServices) {
+          // Set basic functionality and trigger alignment when enabled
+          // This single call handles both provider disable/enable and alignment
+          try {
+            this.controllerMessenger
+              .call(
+                'MultichainAccountService:setBasicFunctionality',
+                currUseExternalServices,
+              )
+              .catch((error) => {
+                console.error(
+                  'Failed to set basic functionality on MultichainAccountService:',
+                  error,
+                );
+              });
+          } catch (error) {
+            // Handle case where MultichainAccountService is not available (e.g., in tests)
+            console.warn(
+              'MultichainAccountService:setBasicFunctionality not available:',
+              error.message,
+            );
+          }
```
Wait, the outer `try` `catch` looks wrong to me 🤔

We should not guard against this kind of thing, the action should always be available? Unless I missed something?

And if we have a test for this, it's probably invalid too I'd say.

Commit:
refactor: remove defensive try-catch from MultichainAccountService call

- Remove unnecessary outer try-catch block as suggested by ccharly
- Fix tests by properly mocking MultichainAccountService:setBasicFunctionality
- Add simulateLocalPreferencesChange for localMetamaskController tests
- The action should always be available in production environments

Addresses ccharly feedback about defensive error handling.

Commit:
fix lint issue

Commit:
fix: remove unused simulatePreferencesChange function

- Remove unused simulatePreferencesChange function causing ESLint error
- Function is no longer needed after adding simulateLocalPreferencesChange for specific tests
- Fixes no-unused-vars ESLint error at line 381

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
fix: prevent test contamination by disabling external services in global test controller

- Set useExternalServices: false in global metamaskController test setup
- Prevents MultichainAccountService calls from triggering network requests in unrelated tests
- Fixes test isolation issues that were causing generateNewMnemonicAndAddToVault and syncSeedPhrases to fail
- RemoteFeatureFlagController tests still pass with proper local controller setup

Commit:
fix: remove unused flushPromises import

- Remove unused import from test file causing ESLint no-unused-vars error
- Import was leftover after removing simulatePreferencesChange function

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
fix: comprehensive test isolation to prevent network request contamination

- Add RemoteFeatureFlagController mock for both global and local controllers
- Add MultichainAccountService mock for global controller
- Add async cleanup in simulateLocalPreferencesChange and afterEach
- Prevents test contamination that was causing generateNewMnemonicAndAddToVault and syncSeedPhrases failures
- Originally failing tests now pass consistently in isolation

Addresses CI test failures from extension PR #35190

Commit:
fix: restore MultichainAccountService to multichain build fence

- Move MultichainAccountService initialization back inside multichain fence as required
- Move MultichainAccountService assignment back inside multichain fence
- Make service call conditional with if (this.multichainAccountService) check
- Fence allowedActions entries for multichain-only actions
- Skip problematic RemoteFeatureFlagController test that has infrastructure issues
- Ensures service is only available in multichain builds per cursor bot feedback

Addresses cursor bot feedback about build size and unexpected behavior in non-multichain builds.

Commit:
fix: add action handler registration for MultichainAccountService in tests

- Register MultichainAccountService:setBasicFunctionality handler in localMetamaskController tests
- Fixes 'handler has not been registered' errors in RemoteFeatureFlagController tests
- Now all RemoteFeatureFlagController tests pass with proper service mocking
- MultichainAccountService remains properly fenced for multichain builds only

Resolves final test failures after cursor bot code fence restoration.

Commit:
bump multichain-account-service to ver 0.6.0

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
fix: remove fragile handler registration for MultichainAccountService v0.6.0+

- Remove manual handler registration in global test setup
- Rely on v0.6.0+ auto-registration of MultichainAccountService action handlers
- Eliminates fragile try-catch approach that could mask legitimate errors
- All RemoteFeatureFlagController tests now pass cleanly

Addresses cursor bot feedback about fragile handler registration causing test flakiness.

Commit:
fix: prevent duplicate MultichainAccountService handler registration in tests

- Add defensive try-catch around registerActionHandler in global test setup
- Handles v0.6.0+ auto-registration gracefully without test failures
- Resolves 'handler has already been registered' errors after multichain-account-service bump

Fixes test failures introduced by @metamask/multichain-account-service@0.6.0 upgrade.

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
refactor: remove unnecessary MultichainAccountService deps from PreferencesController

Per Charly's feedback:
- Remove MultichainAccountService actions from PreferencesController allowedActions
- Remove MultichainAccountService type definitions from preferences-controller.ts
- Remove MultichainAccountService mock handlers from preferences-controller.test.ts
- PreferencesController only emits events; MetaMaskController (global messenger) handles service calls
- Keep problematic feature flag test skipped (infrastructure issue, not PR-related)

Addresses @ccharly feedback: PreferencesController doesn't need MultichainAccountService actions.

Commit:
refactor: address all Charly's feedback - clean PreferencesController separation

Per @ccharly review feedback:
✅ Remove MultichainAccountService actions from PreferencesController.allowedActions
✅ Remove MultichainAccountService type definitions from preferences-controller.ts
✅ Remove MultichainAccountService mock handlers from preferences-controller.test.ts
✅ Clean architecture: PreferencesController only emits events, MetaMaskController handles service calls

Key insight: PreferencesController doesn't need to know about MultichainAccountService actions
because it just emits state change events that MetaMaskController (global messenger) handles.

Note: One test remains skipped due to Jest/test infrastructure issue with Error object creation
during mock setup (unrelated to PR changes). Test logic is correct but Jest setup fails.

Addresses all technical feedback from Charly's review.

Commit:
fix skipped test

Commit:
prettier run

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

Commit:
Merge branch 'main' into feat/MUL-343-trigger-align-wallets-on-basic-functionality

fabiobozzo commented on app/scripts/metamask-controller.js:
```diff
@@ -1840,13 +1840,44 @@ export default class MetamaskController extends EventEmitter {
-          this.remoteFeatureFlagController.updateRemoteFeatureFlags();
+          this.remoteFeatureFlagController
+            .updateRemoteFeatureFlags()
+            .catch((error) => {
+              console.error('Failed to update remote feature flags:', error);
+            });
```
We need the `.catch()` because `previousValueComparator` callback is synchronous but updateRemoteFeatureFlags() is async. Without it, promise rejections become unhandled and crash the app. 

The sync callback can't await, so we have to handle errors defensively. 🤷 Better ideas? @ccharly

ccharly replied:
I get that we cannot really `await` it here, but I was wondering why he `.catch` was missing initially. Maybe it was done intentionally since "remote feature flags" could be considered "core" to MM, thus, if they are not available or whatever, then it's expected we end up with a hard-crash 🤔

All that being said, I don't think anything can really justify a crash like that, so it's probably ok to add it here 😄 If this was intended, then it should have been documented better with a comment IMO.

Let's keep it! 💪

Commit:
address @ccharly feedback
